src/crimson/osd/exceptions.h

```diff
@@ -60,3 +60,525 @@ struct input_output_error : public error {
+  ] (auto&&... args) mutable -> decltype(auto) {
+    if constexpr (std::is_invocable_v<FuncHead, decltype(args)...>) {
+      return std::invoke(std::forward<FuncHead>(head),
+                         std::forward<decltype(args)>(args)...);
+    } else if constexpr (sizeof...(FuncTail) > 0) {
+      using next_composer_t = decltype(composer<FuncTail...>);
+      auto&& next = std::apply<next_composer_t>(composer<FuncTail...>,
+                                                std::move(tail));
+      return std::invoke(std::move(next),
+                         std::forward<decltype(args)>(args)...);
+    } else {
+      static_assert(
+        _impl::always_false<FuncHead>::value, "composition is not exhaustive");
+    }
+  };
+}
+
+template <class... WrappedAllowedErrorsT>
+struct errorator {
+  template <class T>
+  static inline constexpr bool is_error_v = std::is_base_of_v<error_t<T>, T>;
+
+  static_assert((... && is_error_v<WrappedAllowedErrorsT>),
+                "errorator expects presence of ::is_error in all error types");
+
+  template <class... ValuesT>
+  class future : private seastar::future<ValuesT...> {
+    using base_t = seastar::future<ValuesT...>;
+
+    // we need the friendship for the sake of `get_exception() &&` when
+    // `safe_then()` is going to return an errorated future as a result of
+    // chaining. In contrast to `seastar::future`, errorator<T...>::future`
+    // has this member private.
+    template <class ErrorVisitor, class Futurator>
+    friend class maybe_handle_error_t;
+
+    template <class, class = std::void_t<>>
+    struct get_errorator {
+      // generic template for non-errorated things (plain types and
+      // vanilla seastar::future as well).
+      using type = errorator<>;
+    };
+    template <class FutureT>
+    struct get_errorator<FutureT,
+                         std::void_t<typename FutureT::errorator_type>> {
+      using type = typename FutureT::errorator_type;
+    };
+    template <class T>
+    using get_errorator_t = typename get_errorator<T>::type;
+
+    template <class ValueFuncErroratorT, class... ErrorVisitorRetsT>
+    struct make_errorator {
+      // NOP. The generic template.
+    };
+    template <class... ValueFuncWrappedAllowedErrorsT,
+              class    ErrorVisitorRetsHeadT,
+              class... ErrorVisitorRetsTailT>
+    struct make_errorator<errorator<ValueFuncWrappedAllowedErrorsT...>,
+                          ErrorVisitorRetsHeadT,
+                          ErrorVisitorRetsTailT...> {
+      using type = std::conditional_t<
+        is_error_v<ErrorVisitorRetsHeadT>,
+        typename make_errorator<errorator<ValueFuncWrappedAllowedErrorsT...,
+                                          ErrorVisitorRetsHeadT>,
+                                ErrorVisitorRetsTailT...>::type,
+        typename make_errorator<errorator<ValueFuncWrappedAllowedErrorsT...>,
+                                ErrorVisitorRetsTailT...>::type>;
```

Comment:
I think I'm missing something, when would it be valid to call make_errorator with a type which doesn't satisfy is_error_v? and why would it be valid to simply ignore it?

Reply:
For instance:

```cpp
// let's say that `decltype(efut)` is `errorator<enoent>::future<>`
efut.safe_then(
  [] {
    return seastar::now(); // imposes future<>. Visitor affects only error set
  }, enoent::handle([] {
    logger().warn("got enoent; error handling done");
    return seastar::now();
  }));
```

In such case the error visitor's return type is `seastar::future<>` which isn't an error, of course. Those non-error types don't contribute to the new error set, so they aren't handled nor even verified by `make_errorator`. It's rather responsibility of `maybe_handle_error_t::operator()` which does:

```cpp
      using ReturnType = std::result_of_t<ErrorVisitorT(ErrorT)>;
      if constexpr (std::is_assignable_v<decltype(result), ReturnType>) {
        result = std::invoke(std::forward<ErrorVisitorT>(errfunc),
                             ErrorT::from_exception_ptr(std::move(ep)));
      } else if constexpr (!std::is_same_v<ReturnType, void>) {
        // void denotes explicit discarding
        static_assert(_impl::always_false<ReturnType>::value,
                      "return of Error Visitor is not assignable to future");
        // do nothing with `ep`.
      }
```

Reply:
In that case, what happens if a handler could (conditionally) return either of 2 errors or a value?  The user would want in that case to return an errorator<ErrorT, Error2T>::future<Val>.  This machinery would need to merge ErrorT and Error2T into the errorator as well as validate that Val is correct.

Reply:
At the moment an error visitor can return only:

* error types or
* non-error-types which are assignable to the `maybe_handle_error_t::result` .

Generally speaking a `future<T>` (plain or errorated) can be assigned to another `future<T>` only if the former is *not-less-errorated* than the initial one – otherwise the conversion operator is sfinae'd-out. This boils down to that if the `valfunc` returns `seastar::future<T>`, then `errfunc` returning `errorator<new_err>::future<T>` will fail the static assertion for assignability. Adding support for extracting and appending `new_err` to the error set wouldn't be a big issue, I think. At the moment there was simply no handler requiring such thing.

Combining errors and assignable types (by which I mean
```cpp
if constexpr (cond) {
  return less_errorated_future;
} else {
  return new_err::make();
}```)
is supported.

Reply:
Yeah, that makes sense.  I think it'll be worthwhile eventually, but ok to put it off.  I do think you should statically assert that the returned type isn't an errorated future though until then.